Clark's online simulator

clarkwinkelmann opened the pull request:
It looks like everyone had the same idea ahah, here's a fifth online simulator.

Not sure how we want these listed/organized. I have added my link to the README like the previous PR.

My version is a bit different because it's actually implemented as a plugin for the community framework Flarum. CSV links can be shared in posts for instant previews, and the homepage shows user submissions as well as an option for previewing files via drag and drop. I have enabled password and GitHub logins if anyone wants to give it a try.

Since I wasn't sure what the framerate was (I have seen the other discussions on the subject) I made the frame duration customizable in the player.

Tool can be used at https://500led.winkelmann.dev/
Source code is at https://github.com/clarkwinkelmann/flarum-500led

![image](https://user-images.githubusercontent.com/5264300/147632839-c3e080a4-1656-4995-89f6-f465234aaafc.png)

rendermaniac commented:
This is really nice! Thanks for putting this together!

d-albrecht commented:
That really is nit picking on my part because I really like the idea and execution, but the simulator seems to add an additional frame to my animations. I can't really tell if the simulator actually shows one frame more (because the frame number doesn't update when I move the slider while the animation is paused), but at least the counter shows 1501 when there should be exactly 1500 frames in my v1 animations. Just as another random example, Santa Hat is listed with 450 frames but only has 449.

avenzi commented:
I looks like, by default, the website's tree is rendered looking from the positive y-direction. In a statement [here](https://github.com/standupmaths/xmastree2021/issues/21#issuecomment-1002297485), Matt notes that in his video the standard viewing angle will be "down the positive x-axis (where the y axis goes from left to right).  This probably doesn't make much of a difference in most cases, but if anyone's animation assumes that viewing angle, you have to manually rotate the tree by 90 degrees. I only note this because my own animations assume as much :p

clarkwinkelmann commented:
Thanks for the comments!

> the website's tree is rendered looking from the positive y-direction

Good point. I just imported the coordinates into THREE.js and wasn't sure it would be flipped or rotated correctly. Apparently it isn't. I'll try to get this fixed, shouldn't be too complicated. I think I just left the camera on the same axis as the default THREE.js example :hear_no_evil: I see other implementations show the coordinates in 3D space which help with knowing how the tree is actually orientated.

> the simulator seems to add an additional frame to my animations

I'll investigate. I noticed my counter might be a bit confusing because the frame numbers start from zero, so there's at least one bug where the frame total is one more than the actual frames.

I also wasn't able to find whether we are allowed to skip frames in the CSV files? Currently my script reads the lines one by one and assigns the next frame number without looking at the first column.

clarkwinkelmann commented:
I have pushed a fix with some improvements. Now I think my axis are aligned as intended.

I added a toggle to show the axis and a toggle for full screen rendering.

I have improved the player UI so the frame count now updates as the range input is being interacted with.

@d-albrecht can you link to a CSV with the frame count issue? I was not able to replicate.

I have taken a look again but both "santa hat" effects I can find have in fact 450 frames in their CSV. This one has frames 0 to 449 https://github.com/standupmaths/xmastree2021/blob/main/examples/santa_hat_effect.csv

d-albrecht commented:
Okay, with the Santa Hat you're right. I checked the file at least three times, first I thought that the lines didn't match, the next time I opened the raw-view the count matched up and then I checked again and found an off-by-1. Can't tell, why I got confused here several times.

One file where the counter is off is [this one here](https://raw.githubusercontent.com/d-albrecht/XMas-Tree-Explosions/main/v1/explosions3.csv). And I just noticed that the simulator doesn't have a smooth looping. If you pay attention to the red lights in the lower left when the animation starts again, you see the lights flicker for a frame or so.

Btw: being able to load animations per url is really helpful and a feature that I use all the time in other simulators, if you want to further improve your simulator, otherwise you have to download the files first.

clarkwinkelmann commented:
@d-albrecht I have fixed the frame count problem. The issue was with the header line, I had hard-coded the header names in uppercase and your file uses lowercase. This caused the header line to be read as an RGB frame. Thanks for pointing out the issue!

It was already possible to preview an animation from a URL from the forum text editor, but that required being logged in, so probably wasn't obvious and required creating an account. I have now added an input on the homepage to provide a GitHub URL.

I have also added some more error messages in case the CSV file contains unexpected data.

d-albrecht commented:
Okay, so again my fault. (sort of). That wasn't the first simulator that stumbled upon my naming scheme. But by now I adopted the Harvard-format, so although my old files work now, too, my current format doesn't even need this fix anymore.

Was the flickering just a side effect of the "unparsable" header line? All values were set to 0 for this one frame because texts aren't recognizable as numbers?

clarkwinkelmann commented:
Not really your fault, the animation format isn't clearly documented anywhere :sweat_smile: 

Now I implemented it in a way where the header row is optional, and the first column of the first row just needs to be anything except an integer for the first line to be ignored.

I think yes, previously all pixels would be set to black in the invalid frame 0. The value like `r_1` would be converted to `NaN` via `parseInt()`, and then used in a bitwise shift operation `<<` that would convert it to zero.

In my new version if any value becomes `NaN` after `parseInt()`, an error will be thrown during file loading instead of rendering an invalid frame.
